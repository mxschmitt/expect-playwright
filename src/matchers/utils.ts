import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
