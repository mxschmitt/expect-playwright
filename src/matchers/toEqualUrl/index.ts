import { SyncExpectationResult } from "expect/build/types"
import { Page } from "playwright-core"
import { getMessage } from "../utils"

const toEqualUrl: jest.CustomMatcher = async function (
  page: Page,
  expectedUrl: string
): Promise<SyncExpectationResult> {
  const actualUrl = page.url()

  return {
    pass: actualUrl === expectedUrl,
    message: () => getMessage(this, "toEqualUrl", expectedUrl, actualUrl),
  }
}

export default toEqualUrl
