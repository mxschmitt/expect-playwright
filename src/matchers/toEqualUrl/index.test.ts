import toEqualUrl from "."
import { assertSnapshot } from "../tests/utils"

expect.extend({ toEqualUrl })

describe("toEqualUrl", () => {
  const urlPrefix = "http://i-do-not-exist.com"

  beforeAll(async () => {
    await page.route("**/1.html", (route) => {
      route.fulfill({
        body: "123",
        headers: {
          "Content-Type": "text/html",
        },
      })
    })
  })

  afterEach(async () => {
    await page.setContent("")
  })

  it("positive in frame", async () => {
    const myUrl = `${urlPrefix}/1.html`
    await page.setContent(`<iframe src="${myUrl}"></iframe>`)
    const handle = await page.$("iframe")
    const iframe = await handle?.contentFrame()
    await expect(handle).toEqualUrl(myUrl)
    await expect(iframe).toEqualUrl(myUrl)
  })

  it("positive", async () => {
    const myUrl = `${urlPrefix}/1.html`
    await page.goto(myUrl)
    await expect(page).toEqualUrl(myUrl)
  })

  it("negative", async () => {
    await page.goto(`${urlPrefix}/1.html`)
    await assertSnapshot(() => expect(page).toEqualUrl(`${urlPrefix}/2.html`))
  })

  describe("with 'not' usage", () => {
    it("positive in frame", async () => {
      const myUrl = `${urlPrefix}/1.html`
      await page.setContent(`<iframe src="${myUrl}"></iframe>`)
      const handle = await page.$("iframe")
      const iframe = await handle?.contentFrame()
      await expect(handle).not.toEqualUrl("foobar")
      await expect(iframe).not.toEqualUrl("foobar")
    })

    it("positive", async () => {
      await page.goto(`${urlPrefix}/1.html`)
      await expect(page).not.toEqualUrl("foobar")
    })

    it("negative", async () => {
      const myUrl = `${urlPrefix}/1.html`
      await page.goto(myUrl)
      await assertSnapshot(() => expect(page).not.toEqualUrl(myUrl))
    })
  })
})
