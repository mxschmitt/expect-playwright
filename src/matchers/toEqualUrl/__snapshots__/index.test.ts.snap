// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toEqualUrl negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoEqualUrl[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m\\"http://i-do-not-exist.com/[7m2[27m.html\\"[39m
Received: [31m\\"http://i-do-not-exist.com/[7m1[27m.html\\"[39m"
`;

exports[`toEqualUrl with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoEqualUrl[2m([22m[32mexpected[39m[2m)[22m

Expected: not [32m\\"http://i-do-not-exist.com/1.html\\"[39m"
`;
